### DISCUSSION TITLE
Raspberry pi synth concepts

### DISCUSSION THREAD

**Discussion by Gubbledenut** [bazmonk/digigurdy-baz#55] [General]
The basic samperbox code doesn't support multi channel midi, but it does work using a simple 7 segment display and two buttons to switch samples, with no code or config changes. The advantage is that everything can be powered from a single usb source , multiple samples loaded and chosen, output through the inbuilt day to headphones or speaker.making it totally stand/play alone

More recent versions of Samperbox require config adjustments to run, which have so far eluded me to get working
Running Fluidsynth on a pi4 with a simple gui, maybe for testing using a second 1306 oled display and some buttons to allow to select and adjust the levels of the 6 channels as used in bs16i. it was the original setup on android devices, using it's fluidsynth front end version. 
Alternatively this may be an option as a pi "hat" includes 4 buttons, an oled and a better DAC, I imagine bluetooth speaker capability would also be possible through the pi?  https://shop.pimoroni.com/products/pirate-audio-line-out?variant=31189750546515

**Comment by bazmonk** [comment]
It would be outside the scope of this project.  It would be a separate repository at the least…

Here’s my take on it right now:

- Bluetooth audio has unacceptably high latency for playing instruments live over it.  Unless there’s something I’m missing, the delay is noticeable to the point it’s unplayable for me, like trying to talk on a phone where you can hear your own voice with a slight delay.  
- I almost prefer the MIDI connection be over Bluetooth and the sampler/speaker combo having the wires rather than having an audio cable coming out of the gurdy.  With a separate synth, the gurdy itself *can* be wireless.
- The synth/sampler end of a MIDI workflow has to be easily adjustable.  I think this is a big hurdle to building it into the unit.  At the least, I would want to be able to, relatively easily, adjust:
  - The volume of each channel
  - The sound being used on each channel
  - Chorus, reverb, delay, LFO.
To do this needs either a set of hardware controls (knobs, not just buttons) and a dedicated display, a touchscreen, or a web control (at which point you’re back to having a phone/tablet near you).  I really think this kind of control is crucial to it sounding good enough to actually record and work with, and I’m personally not interested in a solution that doesn’t offer it.
- I think micro tuning the channels is a significant difference and I wouldn’t build myself a solution that doesn’t offer it now.  Wouldn’t need to be configurable in the interface, but something I could set it up to do.

I still have some hope for fluidsynth, since it has a webgui, and it’s pretty powerful, and I’ve gotten it to at least work (making sound, not with all the controls I want).  I just haven’t gotten around to looking at samplerbox (the new one).  Even if it worked, though, I’m still essentially trading in my iPad (which is doing a great job of this) for a wire coming out of my digigurdy.

I’m trying to like the idea but I’m not there yet.

Moving this to the discussion section because this isn’t really an issue with the code.

**Reply by Gubbledenut** [reply]
Agree, I just thought you were using bluetooth audio?
The ideal setup for a pi based synth; 
 Power to the pi via a switched USB C connection from a power bank
Power in and Midi Out to teensy from Pi USB socket
Samples loaded onto SD card into adapter and Pi USB socket
Audio either from Pi 3.5mm socket or USB DAC with line out to wired headphones or  powered speaker/amp

Fluidsynth:
Set up via web gui 
 Or iPad using web gui or Touch OSC
Samples as mentioned stored on spare USB socket as SF2 ???

Yes thanks for looking realise it's a seperate concept, do like the idea of like Inca now to just plug the teensy into it and get output.
One day ....

**Comment by bazmonk** [comment]
> Agree, I just thought you were using bluetooth audio?
> 

I’m using Bluetooth MIDI, which works great… I can have it set up on the iPad for sound and then have the gurdy itself either wireless on the built-in battery, or plugged into my laptop for development/testing reasons at the same time.  

For Bluetooth audio to my headphones, both the Pi and the iPad/iPhone have too much latency.  Have you tried one that you liked?
> Audio either from Pi 3.5mm socket or USB DAC with line out to wired headphones or powered speaker/amp
> 
USB DAC for sure, or the Pirate Audio thing add-on board.  A cheap USB one is plug-and-play on a Pi (super-easy to work with) and you could just get any ol’ audio cable, cut it, strip the wires, and solder it onto a jack mounted onto the gurdy body.  I think the USB route would be the cheapest way to get acceptable audio out of it, but I don’t know how much cost matters for your needs.  

I’ve made video game emulators and media players in a couple different forms with the Pi and the built in audio is noticeably not-that-good.  Definitely worth it to go another route.

> Fluidsynth:
> Set up via web gui
> Or iPad using web gui or Touch OSC
> 
I don’t think touch would be worth it unless you want to either build your own interface for it (I don’t want to :-) ), or get a good enough screen to use the web interface on it.  They’re battery guzzlers…

To do this on a smaller/simpler display, I think you’d need at least a jog wheel and a few buttons, sorta like what the MIDIgurdy had, to be able to control it comfortably.  And it’d need a custom interface.

If you went the webgui route, you’d need some backup way to get it on a wifi network.  The Pi can make its own adhoc hotspot network, but switching your phone or tablet’s network just to change the settings is awkward. It would be better if it was connected to your home wifi, but how do you do that if you can’t access it yet?

One idea: build in a single little button (maybe with a simple led light to indicate it) that triggers it to create its own hotspot, you join the digigurdy network on your device, and from there you can set it up to use your wifi network from then on.  That’d give you a way in for new users and the ability to switch wifi networks without having to build a whole extra screen or interface on the gurdy.  It’ll just join the same wifi it did before unless you hit that button.

Design-wise, though… what’s the big advantage now?  If it still requires having some separate device to access the webgui, and that device could just as well also be the synthesizer (like it is today with bs-16i or some program), what’s the compelling reason to involve the Pi at all?  It seems like in the end, all this does is move the audio jack from your device to your gurdy, but it still involves same number of devices.

That’s where my head is at: the easy solution doesn’t offer much of an advantage, and the all-in-one good solution is difficult and would add a good chunk to the overall cost of the thing.
> Samples as mentioned stored on spare USB socket as SF2 ???
> 
If you want but there’s no need.  It can all fit on the SD card no-problem.  A barebones installation for this purpose would have plenty of space on an 8 GB or maybe even a 4 GB SD card.
> Yes thanks for looking realise it's a seperate concept, do like the idea of like Inca now to just plug the teensy into it and get output.
> 
Is that what John showed off in some recent video that he had rigged up?  The red board I think it was?  We should get John’s input.  He’s the one making these for people… so far my main interest is improving the code that a person can use on their DigiGurdy, so if John already has ideas about what he wants to or is going to use on instruments he’s going to make, I’m more likely to try improving that than think up my own solution.

**Comment by Gubbledenut** [comment]
yes it's something John started , and he put me onto the sampler box , and later...so using his modified samples so far on the basic sampler box...just emailed him if he's not following this.
I think loading fluid synth with a preset level options maybe if that's possible, like the "scenes" in bs16i 
but what till John gets back.. I won't distract you more from the initial gig ....

**Comment by bazmonk** [comment]
I don’t mean to dismiss the idea and I’m open to helping where I can.  I’m just chewing on it :-)

**Comment by Gubbledenut** [comment]
I'm trying to get fuildsynth to run blindly on my pi4 using these instructions, but after I've installed everything it runs but then just keeps saying starting fluidsynth server and not going any further????
Also would like to install the Vienna gurdy SF on it, but for now just happy if it will run one channel????
https://www.koopinstruments.com/information/raspberry-pi-midi-synthesizer

**Comment by Gubbledenut** [comment]
Time to start new thread on the tsunami concept. I can't get John's code to compile so hoping you can get a chance to apply your code to the concept... Looking forward to it.

**Comment by Gubbledenut** [comment]
Ok Found out that John's code will only compile if I select a 3.5 board, it won't compile with a 3.6 or 4.1 ??? Further investigation required, I think I have a spare 3.5 that I can set up in the test rig, but not sure I want to run it on my "working" tenor digigurdy with it's partially hardwired 4.1..tried to compile the tsunami test file with no issues for 4.1 ????

**Comment by bazmonk** [comment]
I'll dive back into all this tonight... life caught up with me this
weekend.

**Comment by Gubbledenut** [comment]
Updated to the latest teensyduino 1.57beta1 and now have John's Tsunami code compiles and working on my 4.1 setup...but too hard to change the code to run full tenor keys..will await your version....it sounds rather good...
